codex/blockexchange/swarm.nim

```diff
@@ -0,0 +1,612 @@
+      await self.discovery.removeProvider(peers[i].data.peerId)
+    else:
+      if self.peers.len < self.maxNeighbors:
+        self.addNeighbor(peers[i].data.peerId)
+
+  trace "Swarm neighbors after resample", count = self.peers.len
+
+proc getStalePeers*(self: Swarm, n: int = DefaultRefreshPeerCount): seq[PeerId] =
+  proc cmpStale(x, y: SwarmPeerCtx): int =
+    cmp(x.lastRefresh, y.lastRefresh)
+
+  if self.peers.len == 0:
+    return @[]
+
+  var peers = self.peers.values().toSeq
+  peers.sort(cmpStale)
+  return peers[0 .. min(n, peers.len - 1)].mapIt(it.id).toSeq
+
+proc refreshBlockKnowledge(
+    self: Swarm, peer: PeerId
+): Future[void] {.async: (raises: [CancelledError]).} =
+  # Exchanges knowledge on blocks with peering neighbor
+
+  trace "Asking for block knowledge to peer", peer = peer
+  try:
+    if self.peers[peer].pendingBlockKnowledgeRequest:
+      trace "Pending knowledge update already in progress", peer = peer
+      return
+
+    trace "Setup reply future for block knowledge request"
+
+    self.peers[peer].pendingBlockKnowledgeRequest = true
+
+    # We abuse the want list message to ask for block knowledge.
+    await self.network.request.sendWantList(
+      peer,
+      addresses = @[init(BlockAddress, self.cid)],
+      wantType = WantType.WantHave,
+      full = true,
+    )
+
+    # Ideally we should only update this once we get a reply, and temporarily
+    #  exclude the peer from the refresh loop while it is pending.
+    self.peers[peer].lastRefresh = Moment.now()
+  except KeyError:
+    trace "Cannot refresh update timestamp for dropped peer", peer = peer
+
+proc neighborMaintenanceLoop*(self: Swarm): Future[void] {.async: (raises: []).} =
+  try:
+    trace "Starting neighbor maintenance loop", cid = self.cid
+    while true:
+      # Should check/remove dead peers.
+      if self.peers.len < self.minNeighbors:
+        trace "Too few neighbors, resampling", neighbors = self.peers.len
+        # XXX need backoff when you can't get enough neighbors
+        await self.resample()
+
+      trace "Resampled, neighbors", neighbors = self.peers.len
+
+      let peers = self.getStalePeers()
+      for peer in peers:
+        info "Refreshing block knowledge for peer", peer = peer
+        await self.refreshBlockKnowledge(peer)
+
+      # We should use separate refresh timers per peer and run this in an
+      # event driven fashion.
+      await sleepAsync(1.seconds)
+  except CancelledError:
+    trace "Swarm neighbor maintenance loop cancelled. Exiting."
+
+proc advertiseLoop*(self: Swarm): Future[void] {.async: (raises: []).} =
+  try:
+    while true:
+      trace "Advertising CID", cid = self.cid
+      await self.discovery.provide(self.cid)
+      trace "Avertiser going to sleep"
+      await sleepAsync(DefaultAdvertisementInterval)
+  except CancelledError:
+    trace "Advertisement loop cancelled. Exiting."
+
+proc loadBlockKnowledge*(self: Swarm): Future[void] {.async: (raises: []).} =
```

Comment:
I'm not sure this is needed, why duplicate this if it is already in the blockstore? If you're worried about hitting the repostore too oftne, then we can always put a cache of some sort in front?